### PR TITLE
allow symfony 4.0 for dependencies

### PR DESCRIPTION
**Pull request by ElectricMaxxx**
With this PR, we can require symfony 4.0 on packages where we require yours.

#SymfonyConHackday2017

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
you can/should merge it into you 2.x branch also. I used that one, as we still have bundles with phpunit 5.x.

**Comment by xabbuh** [comment]
Duplicate of #83?

**Comment by ElectricMaxxx** [comment]
yes :-)

**Comment by ElectricMaxxx** [comment]
@xabbuh shall i reopen an merge with the commit on the duplicate?

**Comment by xabbuh** [comment]
@ElectricMaxxx Don't know what @matthiasnoback and @stloyd think about it.

**Comment by matthiasnoback** [comment]
Sorry everyone, I'm not actively maintaining this lib myself. Hoping that @stloyd can help you with it!

**Comment by ElectricMaxxx** [comment]
@xabbuh  @stloyd i reopened that as this one goes on v1.x and i hope i/we get symfony 4 access on v1.x too. You can merge this on into master then later on. I changed the requested stuff from #83 and cherry-pcked his commits.

**Comment by ElectricMaxxx** [comment]
and  ... it seems that id's are not case insensitive anymore : https://github.com/symfony/dependency-injection/commit/5168e40f6358739d2e4229da093a760ba4365d44#diff-bcb414d88ff8da791644f16319363483

**Comment by Nyholm** [comment]
Im very interested to make Symfony 4 support happen. By tomorrow would be perfect. How I can I help?

**Comment by fbourigault** [comment]
I just noticed this PR is against 1.x because of PHPUnit 5.x. You can use 2.x with PHPUnit 5 if you use ^5.7 which provides a forward compatibility layer with namespaced PHPUnit classes. Maybe this project should allow phpunit version which have the forward compatibility layer for a larger compatibility range.

**Comment by ElectricMaxxx** [comment]
So what shouls i do? And who is gonna do the review, merge and release in this repository? 

Btw: shouldn't that be moved closer to the core, when so many depend on it?

**Comment by ElectricMaxxx** [comment]
@fbourigault i do not think so. the 1.x branch requires `^4.0|^5.0` and the 2.x branch `^6.0` so i won't be able to override it.

**Comment by fbourigault** [comment]
Phpunit is a dev dependency. So it won’t be an issue.

**Comment by Soullivaneuh** [comment]
> Sorry everyone, I'm not actively maintaining this lib myself. Hoping that @stloyd can help you with it!

@matthiasnoback I'm also interested to maintain the SF test packages if you need more help. :+1:

File: composer.json [hunk 1]
```diff
@@ -14,10 +14,11 @@
...
     "require": {
         "matthiasnoback/symfony-config-test": "^1.0|^2.0",
-        "symfony/dependency-injection": "^2.3|^3.0",
-        "symfony/config": "^2.3|^3.0",
-        "symfony/yaml": "^2.7|^3.0",
+        "symfony/dependency-injection": "^2.7 || ^3.3 || ^4.0",
+        "symfony/config": "^2.7 || ^3.3 || ^4.0",
+        "symfony/yaml": "^2.7 || ^3.3 || ^4.0",
         "sebastian/exporter": "^1.0|^2.0"
+
```
Comment (stloyd) [inline composer.json]: New line should be removed.

File: composer.json [hunk 2]
```diff
@@ -14,10 +14,11 @@
     ],
     "require": {
         "matthiasnoback/symfony-config-test": "^1.0|^2.0",
-        "symfony/dependency-injection": "^2.3|^3.0",
-        "symfony/config": "^2.3|^3.0",
-        "symfony/yaml": "^2.7|^3.0",
+        "symfony/dependency-injection": "^2.7 || ^3.3 || ^4.0",
+        "symfony/config": "^2.7 || ^3.3 || ^4.0",
+        "symfony/yaml": "^2.7 || ^3.3 || ^4.0",
         "sebastian/exporter": "^1.0|^2.0"
```
Comment (stloyd) [inline composer.json]: Can you also replace single `|` with double `||` everywhere?

**Comment by matthiasnoback** [comment]
@Soullivaneuh That's great, the repo has moved to a new organization. You can ask @Nyholm to add you as a maintainer.

**Comment by ElectricMaxxx** [comment]
when having a testing component/package in symfony i would also like to suggest some stuff we have in [symfony-cmf/testing] (https://github.com/symfony-cmf/testing)

**Comment by ElectricMaxxx** [comment]
@stloyd i changes the stuff, ok for you?

File: .travis.yml [hunk 3]
```diff
@@ -2,36 +2,29 @@ sudo: false
...
+cache:
+  directories:
+    - ~/.composer/cache/files
+
+before_install:
+  - phpenv config-rm xdebug.ini || true
+  - composer self-update
+
+install:
+  - if [ ! -z ${SYMFONY_VERSION+x} ]; then composer require --no-update "symfony/dependency-injection:${SYMFONY_VERSION}"; fi
```
Comment (xabbuh) [inline .travis.yml]: `SYMFONY_VERSION` must be `SYMFONY_DI_VERSION` here

**Comment by Nyholm** [comment]
If this should be merged we need to make sure the test passes. 

See the current travis.yml in master. Could you make the 1.x branch's travis.yml file to look the (approx) same?

**Comment by ElectricMaxxx** [comment]
@Nyholm this is the reason why i proposed to merge on into 1.x and then 1.x into master, to have the same. 
The tests are failing now as i fixed the travis.yml on an hint of @xabbuh but i willl have a look at yours.

**Comment by stloyd** [comment]
TBH. I would drop support for version 1.x at all.

**Comment by ElectricMaxxx** [comment]
oh that travis.yml is quite different

**Comment by Nyholm** [comment]
I think we should drop support as well. But I do believe this PR should be merged as the last change to go in to 1.x

**Comment by ElectricMaxxx** [comment]
@Nyholm you got a hint for me, why i need `minimum-stability: beta` and you do not?

**Comment by stloyd** [comment]
@ElectricMaxxx It's to be able to test against Symfony 4, which is not officially released (should be today :))

**Comment by ElectricMaxxx** [comment]
@stloyd i know, that but the other PR goes through without it.

**Comment by ElectricMaxxx** [comment]
I will squash no worries :-9

File: composer.json [hunk 4]
```diff
@@ -12,15 +12,16 @@
             "homepage": "http://php-and-symfony.matthiasnoback.nl"
         }
     ],
+    "minimum-stability": "beta",
```
Comment (Soullivaneuh) [inline composer.json]: The `prefer-stable` option should be added in this case.

Comment (ElectricMaxxx) [reply composer.json]: done.

Comment (xabbuh) [reply composer.json]: I would not add this setting here, but use `composer config minimum-stability beta` in the Travis CI config instead. Setting `prefer-stable` to `true` would have the effect of not ending up with Symfony 3.4 or 4.0 right now (but you would get 3.3.13 instead).

File: .travis.yml [hunk 5]
```diff
@@ -2,36 +2,29 @@ sudo: false
...
-  allow_failures:
-    - php: hhvm
-
-before_script:
-  - if [ "$TRAVIS_PHP_VERSION" != "hhvm" ]; then phpenv config-rm xdebug.ini; fi
-  - if [ "$SYMFONY_DI_VERSION" != "" ]; then composer require --no-update "symfony/dependency-injection:${SYMFONY_DI_VERSION}"; fi
-  - if [ "$PHPUNIT_VERSION" != "" ]; then composer require --no-update "phpunit/phpunit:${PHPUNIT_VERSION}"; fi
-  - if [ "$deps" = "low" ]; then composer update --prefer-lowest; fi
-  - if [ "$deps" = "" ]; then composer install; fi
+      env: SYMFONY_DI_VERSION=3.3.*
```
Comment (Nyholm) [inline .travis.yml]: We do not need these. We should use `dunglas/symfony-lock` instead. See https://github.com/SymfonyTest/SymfonyDependencyInjectionTest/blob/master/.travis.yml

File: .travis.yml [hunk 6]
```diff
@@ -2,36 +2,29 @@ sudo: false
...
+    - php: 7.1
+      env: SYMFONY_DI_VERSION=4.0.*
+
+cache:
+  directories:
+    - ~/.composer/cache/files
+
+before_install:
+  - phpenv config-rm xdebug.ini || true
+  - composer self-update
```
Comment (Nyholm) [inline .travis.yml]: This is not needed

**Comment by ElectricMaxxx** [comment]
Ahh now it becomes clearer.

File: composer.json [hunk 7]
```diff
@@ -13,14 +13,14 @@
         }
     ],
     "require": {
-        "matthiasnoback/symfony-config-test": "^1.0|^2.0",
-        "symfony/dependency-injection": "^2.3|^3.0",
-        "symfony/config": "^2.3|^3.0",
-        "symfony/yaml": "^2.7|^3.0",
-        "sebastian/exporter": "^1.0|^2.0"
+        "matthiasnoback/symfony-config-test": "^1.0 || ^2.0 || ^3.0",
```
Comment (Nyholm) [inline composer.json]: Are you sure you can use 3.0 of this library?

Comment (ElectricMaxxx) [reply composer.json]: travis hit me faster then you :-)

**Comment by ElectricMaxxx** [comment]
but @Nyholm i got an issue with the last build then.

**Comment by ElectricMaxxx** [comment]
@stloyd should we have a release on symfony-config-test v2.x?

**Comment by Nyholm** [comment]
Hm, right. Because it is only the latest version of config-test that supports sf4. 

Try if you can do quick fix to the 2.x branch in symfony-config-test.

**Comment by ElectricMaxxx** [comment]
@Nyholm `^4.0` is still on 2.x of symfony-config-test, or isn't it?

**Comment by ElectricMaxxx** [comment]
yep, but no tag to grab it: https://github.com/SymfonyTest/SymfonyConfigTest/blob/v2.x/composer.json#L19

**Comment by ElectricMaxxx** [comment]
So i would suggest to tag/release kinda `2.2` there and relaod the build here.

**Comment by ElectricMaxxx** [comment]
I hope that works, cause i am out for family time from now until tonight now.

**Comment by ElectricMaxxx** [comment]
@Soullivaneuh when this one is in and release, can you please restart the build in cs-fixer? this would solve our problems on phpcr-bundle and open symfony 4.0 for CMF :-)

**Comment by Nyholm** [comment]
I just made this: https://github.com/SymfonyTest/SymfonyConfigTest/releases/tag/v2.2.0

This is this the last release in the 2x branch

**Comment by ElectricMaxxx** [comment]
Can you restart the build here please, i got no rights.

**Comment by ElectricMaxxx** [comment]
Yippi.

**Comment by Nyholm** [comment]
Thank you !

**Pull request merged**